fvtest/compilerunittest/p/MemoryReferenceExpansion.cpp

```diff
@@ -0,0 +1,1143 @@
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_EQ(instr, startInstr->getNext());
+    ASSERT_EQ(instr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, instr->getNext());
+
+    ASSERT_EQ(TR::InstOpCode::lwz, instr->getOpCodeValue());
+    ASSERT_EQ(dataReg, instr->getTargetRegister(0));
+    ASSERT_EQ(mr, instr->getMemoryReference());
+    ASSERT_EQ(baseReg, mr->getBaseRegister());
+    ASSERT_EQ(NULL, mr->getIndexRegister());
+    ASSERT_EQ(0x7fff, mr->getOffset());
+}
+
+TEST_F(DISABLED_PPCMemInstructionExpansionTest, smallNegativeDisp) {
+    TR::Register* dataReg = cg()->machine()->getRealRegister(TR::RealRegister::gr0);
+    TR::Register* baseReg = cg()->machine()->getRealRegister(TR::RealRegister::gr1);
+    TR::MemoryReference* mr = new (cg()->trHeapMemory()) TR::MemoryReference(baseReg, -0x8000, 4, cg());
+
+    TR::Node* fakeNode = TR::Node::create(TR::treetop);
+    TR::Instruction* startInstr = generateLabelInstruction(cg(), TR::InstOpCode::label, fakeNode, generateLabelSymbol(cg()));
+    TR::Instruction* instr = generateTrg1MemInstruction(
+        cg(),
+        TR::InstOpCode::lwz,
+        fakeNode,
+        dataReg,
+        mr
+    );
+
+    ASSERT_EQ(instr, instr->expandInstruction());
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_EQ(instr, startInstr->getNext());
+    ASSERT_EQ(instr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, instr->getNext());
+
+    ASSERT_EQ(TR::InstOpCode::lwz, instr->getOpCodeValue());
+    ASSERT_EQ(dataReg, instr->getTargetRegister(0));
+    ASSERT_EQ(mr, instr->getMemoryReference());
+    ASSERT_EQ(baseReg, mr->getBaseRegister());
+    ASSERT_EQ(NULL, mr->getIndexRegister());
+    ASSERT_EQ(-0x8000, mr->getOffset());
+}
+
+TEST_F(DISABLED_PPCMemInstructionExpansionTest, largePositiveDisp) {
+    TR::Register* dataReg = cg()->machine()->getRealRegister(TR::RealRegister::gr0);
+    TR::Register* baseReg = cg()->machine()->getRealRegister(TR::RealRegister::gr1);
+    TR::MemoryReference* mr = new (cg()->trHeapMemory()) TR::MemoryReference(baseReg, 0x8000, 4, cg());
+
+    TR::Node* fakeNode = TR::Node::create(TR::treetop);
+    TR::Instruction* startInstr = generateLabelInstruction(cg(), TR::InstOpCode::label, fakeNode, generateLabelSymbol(cg()));
+    TR::Instruction* instr = generateTrg1MemInstruction(
+        cg(),
+        TR::InstOpCode::lwz,
+        fakeNode,
+        dataReg,
+        mr
+    );
+
+    TR::Instruction* lastInstr = instr->expandInstruction();
+
+    ASSERT_NE(instr, lastInstr);
+    ASSERT_EQ(startInstr, cg()->getFirstInstruction());
+    ASSERT_NE(instr, startInstr->getNext());
+    ASSERT_EQ(lastInstr, cg()->getAppendInstruction());
+    ASSERT_EQ(NULL, lastInstr->getNext());
+
+    ASSERT_TRUE(startInstr->getNext());
+    ASSERT_EQ(TR::Instruction::IsMemSrc1, startInstr->getNext()->getKind());
+    auto stTempInstr = static_cast<TR::PPCMemSrc1Instruction*>(startInstr->getNext());
```

Comment:
This seems fragile as we are assuming the memory reference which cannot encode the large positive displacement will have to emit extra instructions to materialize the memory reference, and we seem to be setting that sequence in stone here via an explicit test. Does this mean it gives us no wiggle room as to what sequence gets expanded?

Reply:
You're right that these tests are decidedly more fragile than many of our other tests and do make assumptions about the sequence we should expand to. This is mostly just the unfortunate reality of trying to unit test an API so deep inside the compiler. The only way to be 100% flexible would be to actually execute the generated code, but that would require running other unrelated parts of the compiler as well (e.g. binary encoding) and would also allow bad implementations to pass the tests by fluke (e.g. we could inadvertently add an instruction that would corrupt a random other volatile register).

The expectation here is that the tests would be updated alongside any update to the instruction sequence we expand to. These tests are really here for 2 primary reasons: firstly, to validate that the implementation expands to the sequence we expect; and secondly to avoid accidentally breaking the expansion code when modifying other parts of the codebase.

Reply:
> The expectation here is that the tests would be updated alongside any update to the instruction sequence we expand to.

That is reasonable to me given we are handling a path not often taken here so the fact that it is fixed is fair. I would rather us have a fixed fully tested path with no surprises than a "more optimal sequence" depending on various factors and have random bugs lurking. Thanks for the exaplanation.